### PR TITLE
Desktop - Server RPC (for stuff like delegate identify ui)

### PR DESCRIPTION
**Pull request by MarcoPolo**
## Context
So far the desktop has been the client and `keybased` has been the server. With the new ability to delegate the identify ui, `keybased` will be the client and desktop will be the server. So we need some new code that will make this pattern simple. That's what this PR introduces. Along with a demo of the identify ui action being printed out in the console

This is a continuation of @maxtaco's ideas on this: https://github.com/keybase/client/pull/1304.

The changes where inspired by wanting to avoid having to manually deal with the start/end functions to call and wanting to be able to simple return values or promises of values from the endpoints.

You should take a look at `react-native/react/native/notifications.desktop.js` first to see how it works. Then you can look at the engine stuff to see how it's implemented.

## Server RPC

A call is made from `keybased` to the desktop to start the `identifyUI` this is called: `'keybase.1.identifyUi.delegateIdentifyUI'`.

The client replies with a sessionId to use for the session. It creates a new endpoint map using the `endpointMapFn` that's passed in as the 4th argument. The state of the session is within that scope, so if you need extra state that should be closed over.

After exchanging messages, `keybased` is finally done and sends `keybase.1.identifyUi.finish` to the desktop. We can listen to that message in the endpoint map like normal, and it will automatically clean up listeners in the callmap for us.

Future PR will include actually spawning the tracker popup window and stuff.

## Types

I also did some type fixing stuff so `.*desktop.js` is no longer being ignored, and this has no flow type errors.

@keybase/react-hackers

### CODE REVIEW

File: react-native/react/engine/index.js [hunk 1]
```diff
@@ -92,6 +94,24 @@ class Engine {
...
+    const cbs = {
+      start: callMap => {
+        this.sessionIDToIncomingCall[sid] = callMap
+        response.result(sid)
+      },
+      end: () => {
+        delete this.sessionIDToIncomingCall[sid]
+      }
+    }
+    this.serverListeners[method](param, cbs)
```
Comment (chrisnojima) [inline react-native/react/engine/index.js]: Should we sanity check serverListeners[method] here

File: react-native/react/native/notifications.desktop.js [hunk 2]
```diff
@@ -21,6 +22,61 @@ export function enableNotifications () {
         console.log('Enabled Notifications')
       }
     })
+
+    // TODO move this somewhere else
```
Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: we should move this to an action

File: react-native/react/native/notifications.desktop.js [hunk 3]
```diff
@@ -21,6 +22,61 @@ export function enableNotifications () {
...
+      finish: (params: {sessionID: number}) => {
+        console.log('finish', params)
+      }
+    }
+
+    const identifyUIServer = new Server(
+      engine,
+      'keybase.1.identifyUi.delegateIdentifyUI',
+      'keybase.1.identifyUi.finish',
+      (params) => { return { keybase: { '1': { identifyUi } } } }
```
Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: remove ()

File: react-native/react/native/notifications.desktop.js [hunk 4]
```diff
@@ -21,6 +22,61 @@ export function enableNotifications () {
...
+    )
+
+    identifyUIServer.listen()
```
Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: is there a reason to have an init + a listen vs just a factory that does this? like will we ever do anything with this object?

File: react-native/react/engine/server.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
+'use strict'
```
Comment (chrisnojima) [inline react-native/react/engine/server.js]: Can we put in some comments about what this is and how it works, very hard to tell

File: react-native/react/engine/server.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        (memo, subkey) => Object.assign(memo, this.parseNextLevel('.', key + separator + subkey, submap[subkey])),
+        {}
+      )
+    } else {
+      // $FlowIssue Computed property keys not supported
+      return {[key]: value}
+    }
+  }
+
+  promisifyResponses (endpoints: FlatServerEndpoints): FlatServerEndpoints {
```
Comment (chrisnojima) [inline react-native/react/engine/server.js]: is this something that should be handled here in this class? why

File: react-native/react/engine/server.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    this.engine = engine
+    this.startMethodName = startMethodName
+
+    this.endpointsFn = (params, end) => {
+      const endpoints = this.parseEndpoint(endpointMapFn(params))
+
+      // End the server when endMethodName is called
+      if (endMethodName) {
+        const originalEndMethod = endpoints[endMethodName]
+        endpoints[endMethodName] = (params) => {
```
Comment (chrisnojima) [inline react-native/react/engine/server.js]: remove ()

File: react-native/react/engine/server.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+          end()
```
Comment (chrisnojima) [inline react-native/react/engine/server.js]: does the order matter? should we call our fn before we call end()?

Comment (MarcoPolo) [inline react-native/react/engine/server.js]: order shouldn't matter, this is the last rpc call we'll get from the service, and all `end()` does is delete the session id from the incoming call map.

File: react-native/react/engine/server.js (hunk 5 above)
Comment (MarcoPolo) [inline react-native/react/engine/server.js]: sure thing!

File: react-native/react/engine/index.js (hunk 1 above)
Comment (MarcoPolo) [inline react-native/react/engine/index.js]: sounds good

File: react-native/react/native/notifications.desktop.js (hunk 4 above)
Comment (MarcoPolo) [inline react-native/react/native/notifications.desktop.js]: You mean just have `new Server` construct the thing and listen at the same time?

Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: i'd prefer a factory that's like createServer(...) that does new + listen. If the pattern is me always making the object and an error to not call listen, and i never will do anything with the object i don't really need the object + 1 method

File: react-native/react/engine/server.js (hunk 6 above)
Comment (MarcoPolo) [inline react-native/react/engine/server.js]: We talked on slack about this and decided to move this and the flatten call map function above into a separate file that will work as call map middlewares.

We'll play around with the idea of using promises for the call maps and if it works well then we'll eventually move exisiting code to use that too.

Comment (chrisnojima) [inline react-native/react/engine/server.js]: :+1:

File: react-native/react/engine/index.js (hunk 1 above)
Comment (MarcoPolo) [inline react-native/react/engine/index.js]: Turns out this code is only called if there is a `this.serverListeners[method]`. See: [here](https://github.com/keybase/client/pull/1318/files#diff-9faed3ed3de893e2f4a99f81e783574eR178)

If there isn't it'll print the existing error of `Unknown incoming RPC`

Comment (chrisnojima) [inline react-native/react/engine/index.js]: k

File: react-native/react/native/notifications.desktop.js (hunk 2 above)
Comment (MarcoPolo) [inline react-native/react/native/notifications.desktop.js]: Will do that in a future PR when I connect to tracker popup

**Comment by MarcoPolo** [comment]
@keybase/react-hackers updated! please TAL!

**Comment by chrisnojima** [comment]
:+1:

**Pull request merged**